libraries/AP_Common/AP_ExpandingArray.h

```diff
@@ -0,0 +1,115 @@
+/*
+   This program is free software: you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation, either version 3 of the License, or
+   (at your option) any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include <AP_Common/AP_Common.h>
+
+template <typename T>
+class AP_ExpandingArray
+{
+public:
+
+    AP_ExpandingArray<T>() :
+        chunk_size(50)
+    {   // create one chunk
+        expand(1);
+    }
+
+    AP_ExpandingArray<T>(uint16_t num_chunks, uint16_t elements_per_chunk) :
+        chunk_size(elements_per_chunk)
+    {   // create requested number of chunks
+        expand(num_chunks);
+    }
+
+    /* Do not allow copies */
+    AP_ExpandingArray<T>(const AP_ExpandingArray<T> &other) = delete;
+    AP_ExpandingArray<T> &operator=(const AP_ExpandingArray<T>&) = delete;
```

Comment:
We may actually want to implement delete for this, since otherwise this will leak memory. We don't normally delete things, but most things aren't manipulating memory the way this is.

Reply:
I'm not sure I understand how we could be leaking memory.  The array may expand to hold everything required but this isn't technically the same as "leaking" which implies the memory is completely lost after use.

So far we don't have a need to delete these objects though so I think it's OK that these definitions that save developers from accidentally creating copies remain.

Reply:
The two operators you overload here block copies and assignment. They don't blocking deleting the object. If you ever allocate then delete one of these arrays all the backing memory the array allocates is forever lost.

Reply:
@WickedShell, ah, I undestand.  OK, I'll add a destructor.

Reply:
destructor added